### PR TITLE
Added K6 templates.

### PR DESCRIPTION
**Pull request by jsclifford**
* Added K6 Templates for easy workflow creation for performance tests with k6.

### CODE REVIEW

**Review by danielle-casella-adams — CHANGES_REQUESTED** [review]
Instead of duplicating templates we've tried to create a single template with the optional items commented out so users can put in or take out what they need.  The manual stress and manual advanced are a great candidate for this as are the manual and k6-ci.  

If you are expecting users to add something, it needs a # TODO:  

We've been pretty explicit about the inputs, so the numbers for inputs (1, 2, 3) should be changed to the environment.  Our environments will be well known, dev, qa, stage, prod, demo, uat so that shouldn't be an issue and if you're worried about mistyping, you can use the map-input action.

File: workflow-templates/im-test-k6-ci.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    container: docker://loadimpact/k6:latest
+
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v1
+    
+    - name: Run k6 test - docker
+      working-directory: tests/k6
+      run: |
+        k6 run ${{env.testFile}} --vus 1 -i 1 --insecure-skip-tls-verify -out JSON="./k6-output.json"
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-ci.yml:24]: I would combine this with the other one and just use commented out items with TODOs to indicate the different options.

File: workflow-templates/im-test-k6-ci.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+name: K6 Automated Load Tester
+
+on: 
+  push:
+    branches: 
+      - master
+      - main
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-ci.yml:7]: During the migration we'll be changing master to main, so you can change this.

File: workflow-templates/im-test-k6-manual-advanced.yml [hunk 3]
```diff
@@ -0,0 +1,66 @@
+name: Manual trigger K6 test - Advanced
+
+on: 
+  workflow_dispatch:
+    inputs:
+      envToTest:
+        description: 'Environment to run test against [[1] dev, [2] qa, [3] stage]'
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-manual-advanced.yml:7]: This should be changed to use the name of the environment.

File: workflow-templates/im-test-k6-manual-advanced.yml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  workflow_dispatch:
+    inputs:
+      envToTest:
+        description: 'Environment to run test against [[1] dev, [2] qa, [3] stage]'   
+        required: true
+        default: '1'
+      testFolderRoot:
+        description: 'Test folder root directory from root of repo'
+        required: true
+        default: "tests/k6"
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-manual-advanced.yml:13]: For this you should add a TODO: for users to verify whether this is correct for their app.  Most of the items where you provide a value for the default should have that TODO.

File: workflow-templates/im-test-k6-manual-advanced.yml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    runs-on: [self-hosted, ubuntu-20.04]
+
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v1
+
+    - name: Setup K6
+      uses: im-open/setup-k6-perf-testing@v1.0.0
+      with:
+        version: 0.33.0
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-manual-advanced.yml:33]: This should have a todo or an ENV at the top asking the user to verify their version.

File: workflow-templates/im-test-k6-manual-advanced.yml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        elif [[ "${{ github.event.inputs.envToTest }}" == "2" ]]; then
+          echo "env: qa"
+          k6_env='qa'
+        elif [[ "${{ github.event.inputs.envToTest }}" == "3" ]]; then
+          echo "env: stage"
+          k6_env='stage'
+        else
+          echo "Invalid option"
+          exit 1
+        fi
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-manual-advanced.yml:50]: This whole thing should be replaced with https://github.com/im-open/map-input-action/pull/3.  Each of the places where this is done should be updated as well.

**Review by danielle-casella-adams — CHANGES_REQUESTED** [review]

File: workflow-templates/im-test-k6-manual-advanced.yml (hunk 4 above)
Comment (jsclifford) [reply workflow-templates/im-test-k6-manual-advanced.yml:13]: Can you suggest an edit to show me the format of the TODO and where to put it?

File: workflow-templates/im-test-k6-ci.yml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+env:
+  testFile: "tests/k6/check-health-simple.js"
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-ci.yml:10]: @jsclifford Here is an example of the todo
```suggestion
  testFile: "" # TODO:  Add the name of the test file, like tests/k6/check-health-simple.js
```

**Comment by jsclifford** [comment]
> Instead of duplicating templates we've tried to create a single template with the optional items commented out so users can put in or take out what they need. The manual stress and manual advanced are a great candidate for this as are the manual and k6-ci.
> 
> If you are expecting users to add something, it needs a # TODO:
> 
> We've been pretty explicit about the inputs, so the numbers for inputs (1, 2, 3) should be changed to the environment. Our environments will be well known, dev, qa, stage, prod, demo, uat so that shouldn't be an issue and if you're worried about mistyping, you can use the map-input action.

I have only two files now.  and made changes with your recommendations.

File: workflow-templates/im-test-k6-ci.yml (hunk 7 above)
Comment (jsclifford) [reply workflow-templates/im-test-k6-ci.yml]: Added recomendation.

File: workflow-templates/im-test-k6-manual-advanced.yml (hunk 6 above)
Comment (jsclifford) [reply workflow-templates/im-test-k6-manual-advanced.yml]: Added as requested

File: workflow-templates/im-test-k6-manual-advanced.yml (hunk 5 above)
Comment (jsclifford) [reply workflow-templates/im-test-k6-manual-advanced.yml]: Added todo.

File: workflow-templates/im-test-k6-manual-advanced.yml (hunk 4 above)
Comment (jsclifford) [reply workflow-templates/im-test-k6-manual-advanced.yml]: added todo

File: workflow-templates/im-test-k6-manual-advanced.yml (hunk 3 above)
Comment (jsclifford) [reply workflow-templates/im-test-k6-manual-advanced.yml]: Updated.

File: workflow-templates/im-test-k6-ci.yml (hunk 2 above)
Comment (jsclifford) [reply workflow-templates/im-test-k6-ci.yml]: Updated

File: workflow-templates/im-test-k6-ci.yml (hunk 1 above)
Comment (jsclifford) [reply workflow-templates/im-test-k6-ci.yml]: Combined files.

File: workflow-templates/im-test-k6-manual.yml [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        required: true
+        default: 'dev'
+      testFolderRoot:
+        description: 'Test folder root directory from root of repo'
+        required: true
+        default: "" # TODO: Verify this is the correct location of your tests.  IE: tests/k6
+      testFile:
+        description: 'K6 test file to run'     
+        required: true
+        default: '' # TODO: Default test file  IE: check-health-simple.js
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-manual.yml:17]:
```suggestion
        default: '' # TODO: Add default test file  IE: check-health-simple.js
```

File: workflow-templates/im-test-k6-manual.yml [hunk 9]
```diff
@@ -0,0 +1,78 @@
+name: Manual trigger K6 test
+
+on: 
+  workflow_dispatch:
+    inputs:
+      envToTest:
+        description: 'Environment to run test against [dev, qa, stage]'   
+        required: true
+        default: 'dev'
+      testFolderRoot:
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-manual.yml:10]: Something I noticed was your variables have kind of mixed casing.  Some use `_` between words and some use pascal casing.  You might want to change to one or the other for consistency.  The templates in this repo use `-` between words.

File: workflow-templates/im-test-k6-manual.yml [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        required: true
+        default: '' # TODO: Default test file  IE: check-health-simple.js
+      k6_extra_params:
+        description: 'K6 command extra paramaters'
+        required: false
+        default: '' # TODO: Set default paramaters to run on every K6 test. IE. --vus 1 -i 1 --insecure-skip-tls-verify -out JSON="./k6-output.json" --out influxdb="http://influxdb:8086/k6" --config "/scripts/config/dev-config.json"
+      k6_version:
+        description: 'K6 version'
+        required: true
+        default: '0.33.0' # TODO: Set default verion of k6
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-manual.yml:25]: In this case, I probably would not provide a value, I'd leave it up to the user so the workflow does not run correctly until someone explicitly puts a value in.  That's how most of the templates are set up.
```suggestion
        default: '' # TODO: Set default version of k6
```

**Review by danielle-casella-adams — CHANGES_REQUESTED** [review]

File: workflow-templates/im-test-k6-ci.yml [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  k6_test:
+    runs-on: [self-hosted, ubuntu-20.04]
+    container: docker://loadimpact/k6:latest
+
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v1
+    
+    - name: Run k6 test - docker
+      working-directory: tests/k6
```
[the remark this answers is not in the document]
Comment (jsclifford) [reply workflow-templates/im-test-k6-ci.yml]: Added todo

File: workflow-templates/im-test-k6-manual.yml (hunk 9 above)
Comment (jsclifford) [reply workflow-templates/im-test-k6-manual.yml]: Updating casing to match repo.

File: workflow-templates/im-test-k6-manual.yml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+      uses: im-open/setup-k6-perf-testing@v1.0.0
+      with:
+        version: ${{ github.event.inputs.k6_version }}
+
+    - name: Validate Environment Input
+      id: validate-env-input
+      uses: im-open/map-input-action@v1.0.1
+      with:
+        input: ${{ github.event.inputs.envToTest }}
+        input_map: "{ \"dev\": [\"dev\", \"Dev\", \"DEV\"], \"qa\": [\"qa\", \"QA\"], \"stage\": [\"stage\", \"Stage\", \"STAGE\"] }"
```
[the remark this answers is not in the document]
Comment (jsclifford) [reply workflow-templates/im-test-k6-manual.yml:49]: Updated for case insensitivity.

File: workflow-templates/im-test-k6-manual.yml [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+on: 
+  workflow_dispatch:
+    inputs:
+      envToTest:
+        description: 'Environment to run test against [dev, qa, stage]'   
+        required: true
+        default: 'dev'
+      testFolderRoot:
+        description: 'Test folder root directory from root of repo'
```
[the remark this answers is not in the document]
Comment (jsclifford) [reply workflow-templates/im-test-k6-manual.yml]: Updated wording.

File: workflow-templates/im-test-k6-ci.properties.json [hunk 14]
```diff
@@ -0,0 +1,5 @@
+{
+    "name": "Test - K6 Workflow for daily CI",
+    "description": "Workflow template for running K6 performance tests with every merge to master",
```
Comment (danielle-casella-adams) [inline workflow-templates/im-test-k6-ci.properties.json:3]: Sorry I missed this one before.
```suggestion
    "description": "Workflow template for running K6 performance tests with every merge to main",
```

Comment (danielle-casella-adams) [reply workflow-templates/im-test-k6-ci.properties.json:3]: I think this one does need to change, since we're running it for the main branch.

**Review by danielle-casella-adams — APPROVED** [review]

**Pull request merged**